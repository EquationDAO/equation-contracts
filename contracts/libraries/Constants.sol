// SPDX-License-Identifier: BUSL-1.1
pragma solidity ^0.8.0;

library Constants {
    uint32 internal constant BASIS_POINTS_DIVISOR = 100_000_000;

    uint16 internal constant ADJUST_FUNDING_RATE_INTERVAL = 1 hours;
    uint16 internal constant SAMPLE_PREMIUM_RATE_INTERVAL = 5 seconds;
    uint16 internal constant REQUIRED_SAMPLE_COUNT = ADJUST_FUNDING_RATE_INTERVAL / SAMPLE_PREMIUM_RATE_INTERVAL;
    /// @dev 8 * (1+2+3+...+720) = 8 * ((1+720) * 720 / 2) = 8 * 259560
    uint32 internal constant PREMIUM_RATE_AVG_DENOMINATOR = 8 * 259560;
    /// @dev RoundingUp(50000 / 8 * Q96 / BASIS_POINTS_DIVISOR) = 4951760157141521099596497
    int256 internal constant PREMIUM_RATE_CLAMP_BOUNDARY_X96 = 4951760157141521099596497; // 0.05% / 8

    uint8 internal constant VERTEX_NUM = 7;
    uint8 internal constant LATEST_VERTEX = VERTEX_NUM - 1;

    uint64 internal constant RISK_BUFFER_FUND_LOCK_PERIOD = 90 days;

    uint256 internal constant Q64 = 1 << 64;
    uint256 internal constant Q96 = 1 << 96;

    bytes32 internal constant ROLE_POSITION_LIQUIDATOR = keccak256("ROLE_POSITION_LIQUIDATOR");
    bytes32 internal constant ROLE_LIQUIDITY_POSITION_LIQUIDATOR = keccak256("ROLE_LIQUIDITY_POSITION_LIQUIDATOR");
}
