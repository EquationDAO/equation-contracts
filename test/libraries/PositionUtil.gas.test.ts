import {ethers} from "hardhat";
import {loadFixture} from "@nomicfoundation/hardhat-network-helpers";
import {expectSnapshotGasCost} from "../shared/snapshotGasCost";
import {DECIMALS_18, DECIMALS_6, SIDE_LONG, SIDE_SHORT, toPriceX96} from "../shared/Constants";

describe("PositionUtil gas tests", () => {
    async function deployFixture() {
        const PositionUtil = await ethers.getContractFactory("PositionUtil");
        const _positionUtil = await PositionUtil.deploy();
        await _positionUtil.deployed();

        const PositionUtilTest = await ethers.getContractFactory("PositionUtilTest", {
            libraries: {
                PositionUtil: _positionUtil.address,
            },
        });
        const positionUtil = await PositionUtilTest.deploy();
        await positionUtil.setGlobalPosition(1, 2, 3, 4);
        return {positionUtil};
    }

    describe("#calculateUnrealizedPnL", () => {
        describe("side is long", () => {
            it("entry price greater than current price", async () => {
                const {positionUtil} = await loadFixture(deployFixture);
                await expectSnapshotGasCost(
                    positionUtil.getGasCostCalculateUnrealizedPnL(
                        SIDE_LONG,
                        10n ** 18n,
                        toPriceX96("1808.234", DECIMALS_18, DECIMALS_6),
                        toPriceX96("1805.234", DECIMALS_18, DECIMALS_6)
                    )
                );
            });

            it("entry price less than current price", async () => {
                const {positionUtil} = await loadFixture(deployFixture);
                await expectSnapshotGasCost(
                    positionUtil.getGasCostCalculateUnrealizedPnL(
                        SIDE_LONG,
                        10n ** 18n,
                        toPriceX96("1805.234", DECIMALS_18, DECIMALS_6),
                        toPriceX96("1808.234", DECIMALS_18, DECIMALS_6)
                    )
                );
            });
        });

        describe("side is short", () => {
            it("entry price greater than current price", async () => {
                const {positionUtil} = await loadFixture(deployFixture);
                await expectSnapshotGasCost(
                    positionUtil.getGasCostCalculateUnrealizedPnL(
                        SIDE_SHORT,
                        10n ** 18n,
                        toPriceX96("1808.234", DECIMALS_18, DECIMALS_6),
                        toPriceX96("1805.234", DECIMALS_18, DECIMALS_6)
                    )
                );
            });

            it("entry price less than current price", async () => {
                const {positionUtil} = await loadFixture(deployFixture);
                await expectSnapshotGasCost(
                    positionUtil.getGasCostCalculateUnrealizedPnL(
                        SIDE_SHORT,
                        10n ** 18n,
                        toPriceX96("1805.234", DECIMALS_18, DECIMALS_6),
                        toPriceX96("1808.234", DECIMALS_18, DECIMALS_6)
                    )
                );
            });
        });
    });

    it("#calculateMaintenanceMargin", async () => {
        const {positionUtil} = await loadFixture(deployFixture);
        const priceX96 = toPriceX96("1808.234", DECIMALS_18, DECIMALS_6);
        await expectSnapshotGasCost(positionUtil.getGasCostCalculateMaintenanceMargin(100, 1, priceX96, 40, 2, 60000));
    });

    describe("#calculateLiquidationPriceX96", () => {
        const margin = 100000000000n;
        const positionSize = 10n ** 20n;
        const positionEntryPriceX96 = toPriceX96("1808.234", DECIMALS_18, DECIMALS_6);
        const liquidationFeeRate = 400000n;
        const tradingFeeRate = 50000n;

        it("side is long", async () => {
            const {positionUtil} = await loadFixture(deployFixture);
            let position = _newPosition();
            position.margin = margin;
            position.size = positionSize;
            position.entryPriceX96 = positionEntryPriceX96;
            await expectSnapshotGasCost(
                positionUtil.getGasCostCalculateLiquidationPriceX96(
                    position,
                    SIDE_LONG,
                    0,
                    liquidationFeeRate,
                    tradingFeeRate,
                    1000000000n
                )
            );
            await expectSnapshotGasCost(
                positionUtil.getGasCostCalculateLiquidationPriceX96(
                    position,
                    SIDE_LONG,
                    -10,
                    liquidationFeeRate,
                    tradingFeeRate,
                    1000000000n
                )
            );
            await expectSnapshotGasCost(
                positionUtil.getGasCostCalculateLiquidationPriceX96(
                    position,
                    SIDE_LONG,
                    0,
                    liquidationFeeRate,
                    tradingFeeRate,
                    100000000001n
                )
            );
            await expectSnapshotGasCost(
                positionUtil.getGasCostCalculateLiquidationPriceX96(
                    position,
                    SIDE_LONG,
                    -10,
                    liquidationFeeRate,
                    tradingFeeRate,
                    100000000001n
                )
            );
        });

        it("side is short", async () => {
            const {positionUtil} = await loadFixture(deployFixture);
            let position = _newPosition();
            position.margin = margin;
            position.size = positionSize;
            position.entryPriceX96 = positionEntryPriceX96;
            await expectSnapshotGasCost(
                positionUtil.getGasCostCalculateLiquidationPriceX96(
                    position,
                    SIDE_SHORT,
                    0,
                    liquidationFeeRate,
                    tradingFeeRate,
                    1000000000n
                )
            );
            await expectSnapshotGasCost(
                positionUtil.getGasCostCalculateLiquidationPriceX96(
                    position,
                    SIDE_SHORT,
                    -10,
                    liquidationFeeRate,
                    tradingFeeRate,
                    1000000000n
                )
            );
            await expectSnapshotGasCost(
                positionUtil.getGasCostCalculateLiquidationPriceX96(
                    position,
                    SIDE_SHORT,
                    0,
                    liquidationFeeRate,
                    tradingFeeRate,
                    100000000001n
                )
            );
            await expectSnapshotGasCost(
                positionUtil.getGasCostCalculateLiquidationPriceX96(
                    position,
                    SIDE_SHORT,
                    -10,
                    liquidationFeeRate,
                    tradingFeeRate,
                    100000000001n
                )
            );
        });
    });
});

function _newPosition() {
    return {
        margin: 0n,
        size: 0n,
        entryPriceX96: 0n,
        entryFundingRateGrowthX96: 0n,
    };
}
