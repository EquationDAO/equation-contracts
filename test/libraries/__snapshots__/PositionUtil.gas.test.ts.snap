// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PositionUtil gas tests #calculateLiquidationPriceX96 side is long 1`] = `5375`;

exports[`PositionUtil gas tests #calculateLiquidationPriceX96 side is long 2`] = `5442`;

exports[`PositionUtil gas tests #calculateLiquidationPriceX96 side is long 3`] = `8005`;

exports[`PositionUtil gas tests #calculateLiquidationPriceX96 side is long 4`] = `9557`;

exports[`PositionUtil gas tests #calculateLiquidationPriceX96 side is short 1`] = `5538`;

exports[`PositionUtil gas tests #calculateLiquidationPriceX96 side is short 2`] = `5605`;

exports[`PositionUtil gas tests #calculateLiquidationPriceX96 side is short 3`] = `8131`;

exports[`PositionUtil gas tests #calculateLiquidationPriceX96 side is short 4`] = `9791`;

exports[`PositionUtil gas tests #calculateMaintenanceMargin 1`] = `351`;

exports[`PositionUtil gas tests #calculateUnrealizedPnL side is long entry price greater than current price 1`] = `362`;

exports[`PositionUtil gas tests #calculateUnrealizedPnL side is long entry price less than current price 1`] = `256`;

exports[`PositionUtil gas tests #calculateUnrealizedPnL side is short entry price greater than current price 1`] = `260`;

exports[`PositionUtil gas tests #calculateUnrealizedPnL side is short entry price less than current price 1`] = `366`;
